In [2]:
import pandas as pd
import numpy as np

##### ***For downloading a file use urlretrieve***

In [3]:
from urllib.request import urlretrieve

In [4]:
italy_covid_url = 'https://gist.githubusercontent.com/aakashns/f6a004fa20c84fec53262f9a8bfee775/raw/f309558b1cf5103424cef58e2ecb8704dcd4d74c/italy-covid-daywise.csv'

urlretrieve(italy_covid_url,'italy-covid-daywise.csv')


('italy-covid-daywise.csv', <http.client.HTTPMessage at 0x26daed32ba0>)

In [5]:
import matplotlib 

In [6]:
import matplotlib.pyplot as plt

In [7]:
import seaborn as sns

In [8]:
covid_df = pd.read_csv('italy-covid-daywise.csv')

##### ***We use 'pd' as pandas function is being on that dataframe***

*The .read_csv is specifically for csv files to read and convert it to pandas dataframe*

In [9]:
type(covid_df)

pandas.core.frame.DataFrame

In [10]:
covid_df

,date,new_cases,new_deaths,new_tests
0,2019-12-31,0.0,0.0,NaN
1,2020-01-01,0.0,0.0,NaN
2,2020-01-02,0.0,0.0,NaN
3,2020-01-03,0.0,0.0,NaN
4,2020-01-04,0.0,0.0,NaN
...,...,...,...,...
243,2020-08-30,1444.0,1.0,53541.0
244,2020-08-31,1365.0,4.0,42583.0
245,2020-09-01,996.0,6.0,54395.0
246,2020-09-02,975.0,8.0,NaN


***dataframe.info() results column wise details such as datatype and count***

In [11]:
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        248 non-null    object 
 1   new_cases   248 non-null    float64
 2   new_deaths  248 non-null    float64
 3   new_tests   135 non-null    float64
dtypes: float64(3), object(1)
memory usage: 7.9+ KB


In [12]:
covid_df.describe()

,new_cases,new_deaths,new_tests
count,248.000000,248.000000,135.000000
mean,1094.818548,143.133065,31699.674074
std,1554.508002,227.105538,11622.209757
min,-148.000000,-31.000000,7841.000000
25%,123.000000,3.000000,25259.000000
50%,342.000000,17.000000,29545.000000
75%,1371.750000,175.250000,37711.000000
max,6557.000000,971.000000,95273.000000


*Column property (arrtibute)of dataframe gives list of column in dataframe*

In [13]:
covid_df.columns

Index(['date', 'new_cases', 'new_deaths', 'new_tests'], dtype='object')

In [14]:
covid_df.shape

(248, 4)

*we can read multiple different format by using read.csv()*

In [15]:
#Pandas Dataframe is similar to this

covid_data_dict = {
    'date' : ['date1','date2','date3','date4','date5','date6',],
    'new_cases': [1444,1365,996,975,1326],
    'new_deaths' : [1,4,6,8,6],
    'new_tests' : [5443,653,2432,6342]
}


##### *Imagine this picture when ever you are going to use pandas_dataframe*

##### ***Retrive Data From Dataframe***

In [16]:
covid_df

,date,new_cases,new_deaths,new_tests
0,2019-12-31,0.0,0.0,NaN
1,2020-01-01,0.0,0.0,NaN
2,2020-01-02,0.0,0.0,NaN
3,2020-01-03,0.0,0.0,NaN
4,2020-01-04,0.0,0.0,NaN
...,...,...,...,...
243,2020-08-30,1444.0,1.0,53541.0
244,2020-08-31,1365.0,4.0,42583.0
245,2020-09-01,996.0,6.0,54395.0
246,2020-09-02,975.0,8.0,NaN


In [17]:
covid_df['new_cases']

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
        ...  
243    1444.0
244    1365.0
245     996.0
246     975.0
247    1326.0
Name: new_cases, Length: 248, dtype: float64

In [18]:
covid_df

,date,new_cases,new_deaths,new_tests
0,2019-12-31,0.0,0.0,NaN
1,2020-01-01,0.0,0.0,NaN
2,2020-01-02,0.0,0.0,NaN
3,2020-01-03,0.0,0.0,NaN
4,2020-01-04,0.0,0.0,NaN
...,...,...,...,...
243,2020-08-30,1444.0,1.0,53541.0
244,2020-08-31,1365.0,4.0,42583.0
245,2020-09-01,996.0,6.0,54395.0
246,2020-09-02,975.0,8.0,NaN


In [19]:
covid_df['new_tests']

0          NaN
1          NaN
2          NaN
3          NaN
4          NaN
        ...   
243    53541.0
244    42583.0
245    54395.0
246        NaN
247        NaN
Name: new_tests, Length: 248, dtype: float64

***Each Column is Represented by data structure called Series.***

In [20]:
type(covid_df['new_cases'])

pandas.core.series.Series

***We Can Access the value in particular column by using 2d array indexing methods***

In [21]:
covid_df['new_cases'][142]

665.0

*Pandas also provide the .at method to dirrectly retirve at specific row & column*

In [22]:
covid_df.at[142,'new_cases']

665.0

In [23]:
covid_df.new_cases

# Same as covid_df['new_cases']

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
        ...  
243    1444.0
244    1365.0
245     996.0
246     975.0
247    1326.0
Name: new_cases, Length: 248, dtype: float64

*We can access some particular columns by creating a dataframe of that particular columns*

In [24]:
cases_df = covid_df[['date','new_cases']]

In [25]:
cases_df
 

,date,new_cases
0,2019-12-31,0.0
1,2020-01-01,0.0
2,2020-01-02,0.0
3,2020-01-03,0.0
4,2020-01-04,0.0
...,...,...
243,2020-08-30,1444.0
244,2020-08-31,1365.0
245,2020-09-01,996.0
246,2020-09-02,975.0


*All the accessing operation result a dataframe pointing to original dataframe so be cautions around using part of dataframe it can alter original dataframe*

*To Over come this situation we can use .copy() method to make a view of that particular columns of that dataframe*

In [26]:
cases_df2 = covid_df.copy()
cases_df2

,date,new_cases,new_deaths,new_tests
0,2019-12-31,0.0,0.0,NaN
1,2020-01-01,0.0,0.0,NaN
2,2020-01-02,0.0,0.0,NaN
3,2020-01-03,0.0,0.0,NaN
4,2020-01-04,0.0,0.0,NaN
...,...,...,...,...
243,2020-08-30,1444.0,1.0,53541.0
244,2020-08-31,1365.0,4.0,42583.0
245,2020-09-01,996.0,6.0,54395.0
246,2020-09-02,975.0,8.0,NaN


##### *.loc[index] is used to retrive particular row or multiple rows from dataframe*

In [27]:
covid_df.loc[142]

date          2020-05-21
new_cases          665.0
new_deaths         161.0
new_tests        40644.0
Name: 142, dtype: object

*We Can view few rows from dataframe by using .head(no_of_rows) and .tail(no_of_rows)*

In [28]:
covid_df.head(5)

,date,new_cases,new_deaths,new_tests
0,2019-12-31,0.0,0.0,NaN
1,2020-01-01,0.0,0.0,NaN
2,2020-01-02,0.0,0.0,NaN
3,2020-01-03,0.0,0.0,NaN
4,2020-01-04,0.0,0.0,NaN


In [29]:
covid_df.tail(3)

,date,new_cases,new_deaths,new_tests
245,2020-09-01,996.0,6.0,54395.0
246,2020-09-02,975.0,8.0,NaN
247,2020-09-03,1326.0,6.0,NaN


In [30]:
covid_df.at[0,'new_tests']

nan

Italy Started sharing new_tests data after 19th april 2020 till then 935310 test have conducted

*We can retrive the index of the row from where the number is started where before it the data is represented in form of NaN*

In [31]:
covid_df.new_tests.first_valid_index()

111

##### ***We Can Also use python slicing to slice between rows***

In [32]:
covid_df[45:89]

,date,new_cases,new_deaths,new_tests
45,2020-02-14,0.0,0.0,NaN
46,2020-02-15,0.0,0.0,NaN
47,2020-02-16,0.0,0.0,NaN
48,2020-02-17,0.0,0.0,NaN
49,2020-02-18,0.0,0.0,NaN
50,2020-02-19,0.0,0.0,NaN
51,2020-02-20,0.0,0.0,NaN
52,2020-02-21,0.0,0.0,NaN
53,2020-02-22,14.0,0.0,NaN
54,2020-02-23,62.0,2.0,NaN


In [33]:
covid_df.sample(25)

,date,new_cases,new_deaths,new_tests
82,2020-03-22,6557.0,795.0,NaN
100,2020-04-09,3836.0,540.0,NaN
150,2020-05-29,593.0,70.0,38233.0
173,2020-06-21,264.0,49.0,24581.0
148,2020-05-27,397.0,78.0,37299.0
34,2020-02-03,0.0,0.0,NaN
108,2020-04-17,3786.0,525.0,NaN
5,2020-01-05,0.0,0.0,NaN
80,2020-03-20,5322.0,429.0,NaN
63,2020-03-03,347.0,17.0,NaN


##### ***Summary***

 1:- covid_df['column name'] -> access column as series using column name

 2:- new_cases[243] -> access particular row using row index

 3:- covid_df.at[243,'new_cases'] ->retriving a single value from the data frame

 4:- covid_df.copy() -> Makes a copy of dataframe which we not affect the orignal dataframe

 5:-covid_df.loc() -> same as column['row index'] 
            Retriving a row or range of row from dataframe

 6:-covid_df.head(n) -> access first n rows from dataframe

 7:-covid_df.tail(n) -> access last n rows from dataframe

 8:-covid_df.sample(n) -> access random tuple from dataframe to check the distribution of data.

 9:-covid_df.new_tests.first_valid_index -> this is a type of code used for finding first non-empty index in a 'series' 




##### ***Ananlyzing Data from data frames***

Q1.What is the total number of reported cases and deaths related to covid-19 in italy?

In [37]:
total_cases = covid_df['new_cases'].sum()
print(total_cases)

271515.0


In [39]:
total_deaths = covid_df['new_deaths'].sum()
print(total_deaths)

35497.0


In [51]:
death_rate = (total_deaths/total_cases)*100

print("The overall death rate in Italy is {:.2f} %".format(death_rate))

The overall death rate in Italy is 13.07 %


Q2.What is the overall number of test conducted?.A total of 935310 test were conducted before daily test numbers were beign reported.